src/charm.py

```diff
@@ -212,9 +209,36 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:
             self._set_status(Status.SERVICE_UNHEALTHY)
             return
 
+        # in case server was erroneously removed from the quorum
+        if not self.state.stale_quorum and not self.quorum_manager.server_in_quorum:
+            self._set_status(Status.SERVICE_NOT_QUORUM)
+            return
+
         self._set_status(Status.ACTIVE)
 
-    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
+    def _on_cluster_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Handler for `relation_departed` events."""
+        # is related to issue found in https://bugs.launchpad.net/juju/+bug/2053055
+        # likely due to a controller upgrade or a cloud maintenance with machines being reshuffled
+        # periodically, juju would emit a LeaderElected event, and would return no peer units
+        # the leader would then remove all other units from the quorum, which when restarted, would fail
+        if not event.departing_unit:
+            return
+
+        departing_server_id = (
+            int(event.departing_unit.name.split("/")[1]) + 1
+        )  # server-ids must be positive integers
+
+        try:
+            self.quorum_manager.client.remove_members(members=[f"server.{departing_server_id}"])
```

Comment:
Every unit gets the departing event, so in this case, it's ran on every unit.
If another unit got the event first, Kazoo will raise the 'ReconfigInProcessError', or if it was already done, we'll get 'BadVersionError', both indicating another unit has covered it.

Reply:
We don't run on the Juju leader only, because that unit may also be going down, and may not get the event for the other units.

For example:
```
0, 1, 2*
Remove 2* and 1
2* gets 2 departing, removes itself, dies
0 gets 2 departing, is not leader 
0 gets 1 departing, is not leader
0 gets elected leader 0*
```

By this point, we've missed the chance for the Juju leader to remove anything.

Reply:
Oh, I see. I was hoping that, when scaling down the leader, the leader elected hook would be running before the departing ones. I mean: it would make sense to me that first I swap leader, then I tear down the unit :man_shrugging: 

Anyhow, This is fine, as long as the racing condition is taken care of. Resolving the comment